articles/react-pwa/index.md

```diff
@@ -0,0 +1,508 @@
+|--- logo.svg
+.gitignore
+package.json
+README.md
+yarn.lock
+```
+Our React code will be in the `src` folder. Inside this folder, 
+delete the `logo. svg` and remove everything in both `App.css` and `App.js` files.
+
+
+### Creating UI components
+Using React means that we will have to break our application into components that represent the 
+logical parts of the UI.
+We shall have four functional components.
+1. App component
+2. Form component
+3. Todo component
+4. TodoList component
+
+### Form component
+This is the component that will be used to add new todos.
+In your `src/` folder, create a new folder named `components`. Inside the `components` folder, you need to create a file named `Form.js`. 
+This is the code for the *Form.js*:
+
+```javascript
+import React,{useState} from 'react';
+import {v4 as uuidv4} from 'uuid';
+
+const Form =({addTodo})=>{
+    
+    const [inputText,setInputText]=useState("");
+
+    const inputTextHandler=(e)=>{
+        setInputText(e.target.value);
+    }
+    const submitTodoHandler=(e)=>{
+        e.preventDefault();
+        if(!inputText){
+            return
+        }
+        addTodo({text:inputText,completed:false,id:uuidv4()});
+        setInputText("");
+    }
+
+    return(
+        <form onSubmit={submitTodoHandler}>
+            <input type="text" value={inputText} placeholder="Add Todo" onChange={inputTextHandler}/>
+            <button type="submit" className="add-btn">
+                Add
+            </button>
+        </form>
+    );
+};
+
+export default Form;
+```
+
+The code above uses functional components alongside React hooks. Hooks are functions that let you have state and other lifecycle patterns from function components.
+We are using the `useState` hook to return two values in an array. The first value is the state, and the second value is the function to update the state. We have named the destructured values as `inputText` and `setInputText`.
+Inside the form, we are binding the input tag to an event. When the state of the form changes, The method `inputTextHandler` is invoked and updates the state to the current input value by calling the `setInputText` function.
+The form is a controlled component that will track the `inputText`. The `submitTodoHandler` is bound on the form when a button is clicked that will invoke the `addTodo` function. We passed the `addTodo` down as props from the `App` component. The `setInputText` inside the `submitTodoHandler` will simply reset the form.
+We now need to add this component in our parent `App` component and pass the function `addTodo`.
+Below is the `App.js` file.
+```javascript
+import React,{useState} from 'react';
+
+import Form from './components/Form';
+
+const App=()=>{
+
+  const addTodo= todo=>{
+          const newTodos=[todo,...todos];
+           setTodos(newTodos);
+       }
+  
+  return (
+    <div>
+      <h1>TODO PWA</h1>
+      <Form addTodo={addTodo}/>
+    </div>
+  );
+}
+
+export default App;
+```
```

Comment:
I think you can remove this snippet, move to the TodoList component, and Introduce App.js in that section.

Reply:
You mean re implementation by code refactoring, or the explanation workflow?